### PR TITLE
Adding state_dicts test to adam_test

### LINKED ISSUE #116507: Add testing regarding SparseAdam state_dicts
### 🚀 The feature, motivation and pitch

SparseAdam, unlike the other optimizers, does not test any state_dict functionality. All other optims call _test_state_dict, but due to SparseAdam's only handling sparse gradients, it does not easily fall into the architecture that already exists.

We should consider giving it its own closure that would update parameters like in `rosenbrock` and creating special case tests for it.

### Alternatives

Leave SparseAdam state_dict handling untested

### Additional context

_No response_

cc @vincentqb @jbschlosser @albanD @crcrpar @mruberry @ZainRizvi

### PR DESCRIPTION
**Pull request by omastey**
Fixes #116507

### CODE REVIEW

**Review by janeyx99 — CHANGES_REQUESTED** [review]
Ah, we have our optimizer tests in TestOptimRenewed in test/test_optim.py and not in this file which looks outdated. The test case you added also does not address all the concerns in the issue.

What I meant by the issue is to go through the `state_dict` tests in TestOptimRenewed where `SparseAdam` is skipped and to add tests for SparseAdam. SparseAdam requires special handling as it only takes in sparse grads but dense weights.
